File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/SearchSongFilter.kt
package com.bockmedia.console.domain.model

import com.bockmedia.console.data.api.dto.SearchHit
import com.bockmedia.console.data.api.dto.SearchResponse

/** Keep prefix/acronym song matches; drop album tracklist rows and soundtrack suffix titles. */
fun filterSearchSongHits(query: String, songs: List<SearchHit>): List<SearchHit> {
    val q = query.trim()
    if (q.length < 2) return emptyList()
    return songs.filter { hit ->
        matchesSearchSongTitle(q, hit.title, hit.album, hit.artist, hit.genre)
    }
}

internal fun matchesSearchSongTitle(
    q: String,
    title: String?,
    album: String?,
    artist: String? = null,
    genre: String? = null,
): Boolean {
    if (SearchPrefixMatch.fieldMatchesQuery(q, artist)) return true
    if (genreMatchesQuery(q, genre)) return true
    if (!SearchPrefixMatch.fieldMatchesQuery(q, title)) return false
    val al = album?.lowercase()?.trim().orEmpty()
    val tl = title?.lowercase()?.trim().orEmpty()
    for (sep in listOf(" - from ", " – from ")) {
        val idx = tl.indexOf(sep)
        if (idx >= 0) {
            val primary = tl.substring(0, idx)
            val suffix = tl.substring(idx + sep.length)
            if (SearchPrefixMatch.fieldMatchesQuery(q, al) &&
                !SearchPrefixMatch.fieldMatchesQuery(q, primary) &&
                SearchPrefixMatch.fieldMatchesQuery(q, suffix)
            ) {
                return false
            }
        }
    }
    val dash = tl.indexOf(" - ")
    if (dash >= 0 && SearchPrefixMatch.fieldMatchesQuery(q, al)) {
        val primary = tl.substring(0, dash)
        val suffix = tl.substring(dash + 3)
        if (!SearchPrefixMatch.fieldMatchesQuery(q, primary) &&
            SearchPrefixMatch.fieldMatchesQuery(q, suffix)
        ) {
            return false
        }
    }
    return true
}

fun SearchResponse.hasAnyResults(): Boolean =
    songs.isNotEmpty() || artists.isNotEmpty() || albums.isNotEmpty() ||
        playlists.isNotEmpty() || genres.isNotEmpty() || smartPlaylists.isNotEmpty()

/** Client-side narrow when server returns empty but a shorter prefix still matched. */
fun SearchResponse.narrowToQuery(query: String): SearchResponse {
    val q = query.trim()
    return copy(
        songs = songs.filter { matchesSearchSongTitle(q, it.title, it.album, it.artist, it.genre) },
        artists = artists.filter { SearchPrefixMatch.fieldMatchesQuery(q, it.name) },
        albums = albums.filter {
            SearchPrefixMatch.fieldMatchesQuery(q, it.name) ||
                SearchPrefixMatch.fieldMatchesQuery(q, it.artist)
        },
        playlists = playlists.filter { SearchPrefixMatch.fieldMatchesQuery(q, it.name) },
        smartPlaylists = smartPlaylists.filter { SearchPrefixMatch.fieldMatchesQuery(q, it.name) },
        genres = genres.filter { genreMatchesQuery(q, it.name) },
    )
}

/** Genre bucket: prefix match plus substring on folded text (french ↔ Français). */
internal fun genreMatchesQuery(q: String, name: String?): Boolean {
    if (SearchPrefixMatch.fieldMatchesQuery(q, name)) return true
    val qc = SearchPrefixMatch.compact(q)
    val nc = SearchPrefixMatch.compact(name ?: "")
    return qc.length >= 2 && nc.contains(qc)
}

File: ios/BockMedia/Domain/SearchSongFilter.swift
import Foundation

enum SearchSongFilter {
    static func filter(query: String, songs: [SearchHit]) -> [SearchHit] {
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard q.count >= 2 else { return [] }
        return songs.filter { matchesSearchSongTitle(q: q, title: $0.title, album: $0.album, artist: $0.artist) }
    }

    static func matchesSearchSongTitle(q: String, title: String?, album: String?, artist: String? = nil) -> Bool {
        if SearchPrefixMatch.fieldMatchesQuery(q: q, text: artist) { return true }
        guard SearchPrefixMatch.fieldMatchesQuery(q: q, text: title) else { return false }
        let tl = title?.lowercased().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let al = album?.lowercased().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        for sep in [" - from ", " – from "] {
            if let idx = tl.range(of: sep)?.lowerBound {
                let primary = String(tl[..<idx])
                let suffix = String(tl[idx...].dropFirst(sep.count))
                if SearchPrefixMatch.fieldMatchesQuery(q: q, text: al),
                   !SearchPrefixMatch.fieldMatchesQuery(q: q, text: primary),
                   SearchPrefixMatch.fieldMatchesQuery(q: q, text: suffix) {
                    return false
                }
            }
        }
        if let dash = tl.range(of: " - ") {
            let primary = String(tl[..<dash.lowerBound])
            let suffix = String(tl[dash.upperBound...])
            if SearchPrefixMatch.fieldMatchesQuery(q: q, text: al),
               !SearchPrefixMatch.fieldMatchesQuery(q: q, text: primary),
               SearchPrefixMatch.fieldMatchesQuery(q: q, text: suffix) {
                return false
            }
        }
        return true
    }
}

enum SearchPrefixMatch {
    static func compact(_ s: String) -> String {
        s.lowercased().replacingOccurrences(of: "[^a-z0-9]", with: "", options: .regularExpression)
    }

    static func acronymCompact(_ text: String) -> String {
        let parts = text.split { $0 == "." || $0.isWhitespace }.map(String.init).filter { !$0.isEmpty }
        if parts.count > 1 {
            return parts.map { String($0.prefix(1)).lowercased() }.joined()
        }
        let words = text.split { !$0.isLetter && !$0.isNumber }.map(String.init).filter { !$0.isEmpty }
        if words.count > 1 {
            return words.map { String($0.prefix(1)).lowercased() }.joined()
        }
        return compact(text)
    }

    static func fieldMatchesQuery(q: String, text: String?) -> Bool {
        let qc = compact(q)
        guard !qc.isEmpty, let text, !text.isEmpty else { return false }
        let tc = compact(text)
        if tc.hasPrefix(qc) { return true }
        let ac = acronymCompact(text)
        if ac.hasPrefix(qc) { return true }
        let words = text.split { !$0.isLetter && !$0.isNumber }.map(String.init).filter { !$0.isEmpty }
        for word in words {
            if compact(word).hasPrefix(qc) { return true }
            if acronymCompact(word).hasPrefix(qc) { return true }
        }
        return false
    }
}
